x-pack/plugins/alerting/server/lib/wrap_scoped_cluster_client.ts

```diff
@@ -0,0 +1,169 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import {
+  TransportRequestOptions,
+  TransportResult,
+  TransportRequestOptionsWithMeta,
+  TransportRequestOptionsWithOutMeta,
+} from '@elastic/elasticsearch';
+import type {
+  SearchRequest,
+  SearchResponse,
+  AggregateName,
+} from '@elastic/elasticsearch/lib/api/types';
+import type {
+  SearchRequest as SearchRequestWithBody,
+  AggregationsAggregate,
+} from '@elastic/elasticsearch/lib/api/typesWithBodyKey';
+import { IScopedClusterClient, ElasticsearchClient, Logger } from 'src/core/server';
+import { ElasticsearchClientWithChild, RuleExecutionMetrics } from '../types';
+import { Alert as Rule } from '../types';
+
+type RuleInfo = Pick<Rule, 'name' | 'alertTypeId' | 'id'>;
+interface WrapScopedClusterClientFactoryOpts {
+  scopedClusterClient: IScopedClusterClient;
+  rule: RuleInfo;
+  logger: Logger;
+}
+
+type WrapScopedClusterClientOpts = WrapScopedClusterClientFactoryOpts & {
+  logMetricsFn: LogSearchMetricsFn;
+};
+
+type WrapEsClientOpts = Omit<WrapScopedClusterClientOpts, 'scopedClusterClient'> & {
+  esClient: ElasticsearchClient;
+};
+
+interface LogSearchMetricsOpts {
+  queryDuration: number;
+  searchDuration: number;
+}
+type LogSearchMetricsFn = (metrics: LogSearchMetricsOpts) => void;
+
+export function createWrappedScopedClusterClientFactory(opts: WrapScopedClusterClientFactoryOpts) {
+  let numQueries: number = 0;
+  let totalQueryDurationMs: number = 0;
+  let totalSearchDurationMs: number = 0;
+
+  function logMetrics(metrics: LogSearchMetricsOpts) {
+    numQueries++;
+    totalQueryDurationMs += metrics.queryDuration;
+    totalSearchDurationMs += metrics.searchDuration;
+  }
+
+  const wrappedClient = wrapScopedClusterClient({ ...opts, logMetricsFn: logMetrics });
+
+  return {
+    client: () => wrappedClient,
+    getMetrics: (): RuleExecutionMetrics => {
+      return {
+        totalQueryDurationMs,
+        totalSearchDurationMs,
+        numQueries,
+      };
+    },
+  };
+}
+
+function wrapScopedClusterClient(opts: WrapScopedClusterClientOpts): IScopedClusterClient {
+  const { scopedClusterClient, ...rest } = opts;
+  return {
+    asInternalUser: wrapEsClient({
```

Comment:
I wonder if some day we'll want these counts for internal / current separated.  I think we can wait till we see the need, don't think we'd likely be able to tell yet.